crates/sui-core/src/authority_active/gossip/node_sync.rs

```diff
@@ -0,0 +1,324 @@
+use super::{Follower, FollowerDigestHandler};
+use crate::{
+    authority::AuthorityState, authority_aggregator::AuthorityAggregator,
+    authority_client::AuthorityAPI, safe_client::SafeClient,
+};
+use async_trait::async_trait;
+
+use std::collections::{btree_map, hash_map, BTreeMap, HashMap};
+use sui_storage::pending_store::PendingStore;
+use sui_types::{
+    base_types::{AuthorityName, ExecutionDigests, TransactionDigest, TransactionEffectsDigest},
+    committee::{Committee, StakeUnit},
+    error::{SuiError, SuiResult},
+    messages::TransactionInfoRequest,
+};
+
+use std::ops::Deref;
+use std::sync::Arc;
+
+use tokio::sync::{
+    broadcast,
+    mpsc::{channel, Receiver, Sender},
+    Mutex, Semaphore,
+};
+use tokio::task::JoinHandle;
+
+use tracing::{error, info, trace, warn};
+
+const NODE_SYNC_QUEUE_LEN: usize = 500;
+
+// Sync up to 20 certificates concurrently.
+const MAX_NODE_SYNC_CONCURRENCY: usize = 20;
+
+/// EffectsStakeMap tracks which effects digests have been attested by a quorum of validators and
+/// are thus final.
+struct EffectsStakeMap {
+    /// Keep track of how much stake has voted for a given effects digest
+    /// any entry in this map with >2f+1 stake can be sequenced locally and
+    /// removed from the map.
+    effects_stake_map: HashMap<ExecutionDigests, StakeUnit>,
+    /// Keep track of stake votes per validator - needed to double check the total stored in
+    /// effects_stake_map, which can otherwise be corrupted by byzantine double-voting.
+    effects_vote_map: HashMap<(ExecutionDigests, AuthorityName), StakeUnit>,
```

Comment:
This can be simplified to one hashmap like `HashMap<ExecutionDigests, Vec<(AuthorityName, StakeUnit)>`, or even `HashMap<ExecutionDigests, Vec<AuthorityName>`   
Also, is there a reason to store the entire ExecutionDigests as the key, or would it be possible to use more specifically the `TransactionEffectsDigest`?

Reply:
I don't want to use a Vec because then we have to do a linear search to check if the validator is in the vec. I could use nested HashMaps but I find nested structures lead to uglier code most of the time, and that would likely be slower anyway due to more pointer chasing.

As far as the key, a byzantine validator could give us two ExecutionDigests of the form (T1, E1), (T2, E1), so we can't assume that the EffectsDigest is unique per tx.

Reply:
I agree with Laura's take on this. it would be a bit more ideal to store the nested structure, probably as a Hashmap instead of a Vec.

I would use just the EffectsDigest as a key as well to make the keys smaller. In fact all ExecutionDigests can be folded down to just an EffectsDigest and the Effects structure points to the TransactionData structure.

Reply:
Ok, turns out there's a good reason to do it your way so my personal taste is irrelevant, which is that I forgot to garbage collect the votes after processing the cert, and that you can address the map with digest only.